### PR TITLE
Merge pytest-xdist latest changes + correct lint errors in nTop fork

### PR DESCRIPTION
**Pull request by ciaconet**
### Summary
Merge latest changes to upstream and correct lint errors in fork.

### Testing
```
python -m mypy src --ignore-missing-imports --strict

python -m ruff check
```

### CODE REVIEW

File: src/xdist/dsession.py [hunk 1]
```diff
@@ -287,7 +296,9 @@ def worker_workerfinished(self, node: WorkerController) -> None:
                 try:
                     self.prepare_for_reschedule()
                 except Exception as e:
-                    self.shouldstop = f"Exception caught during preparation for rescheduling. Giving up.\n{''.join(traceback.format_exception(e))}"
+                    msg = ("Exception caught during preparation for rescheduling. Giving up."
```
Comment (Heroico) [inline src/xdist/dsession.py:298]: 😢

File: src/xdist/scheduler/load.py [hunk 2]
```diff
@@ -62,6 +63,9 @@ def __init__(self, config: pytest.Config, log: Producer | None = None) -> None:
         self.node2collection: dict[WorkerController, list[str]] = {}
         self.node2pending: dict[WorkerController, list[int]] = {}
         self.pending: list[int] = []
+        self.do_resched: bool = False
```
Comment (Heroico) [inline src/xdist/scheduler/load.py]: Can we avoid modifying the vanilla schedulers?
e.g. checking for member definition in `Dession`, e.g. `hasattr(self.sched, 'do_resched')` there

**Review by Heroico — APPROVED** [review]
I don't like what the linter demands but don't want to fight the upstream either.

File: src/xdist/scheduler/load.py (hunk 2 above)
Comment (ciaconet) [reply src/xdist/scheduler/load.py]: Let me see if that works! I'd also like to not modify protocol/existing schedulers

File: src/xdist/workermanage.py [hunk 3]
```diff
@@ -83,7 +82,7 @@ def rsync_roots(self, gateway: execnet.Gateway) -> None:
     def setup_nodes(
         self,
         putevent: Callable[[tuple[str, dict[str, Any]]], None],
-        max_nodes: Optional[int] = None    
+        max_nodes: int | None = None
```
Comment (Heroico) [inline src/xdist/workermanage.py:85]: I wish we could use `Optional` 😢

**Review by Heroico — APPROVED** [review]
Thanks, this should help communications with the upstream.

**Comment by mlopez-ntop** [comment]
Locally I got the following. `xdist` is from the upstream right?
```(dev-11) C:\Users\mattlopez\Documents\pytest-xdist>python -m mypy src --ignore-missing-imports --strict
src\xdist\looponfail.py:169: error: Incompatible types in assignment (expression has type "ChannelFileWrite", variable has type "TextIO")  [assignment]
src\xdist\plugin.py:34: error: Returning Any from function declared to return "int"  [no-any-return]
src\xdist\plugin.py:36: error: Module "os" has no attribute "sched_getaffinity"  [attr-defined]
Found 3 errors in 2 files (checked 18 source files)```

**Comment by ciaconet** [comment]
> Locally I got the following. `xdist` is from the upstream right?
> 
> ```
> src\xdist\looponfail.py:169: error: Incompatible types in assignment (expression has type "ChannelFileWrite", variable has type "TextIO")  [assignment]
> src\xdist\plugin.py:34: error: Returning Any from function declared to return "int"  [no-any-return]
> src\xdist\plugin.py:36: error: Module "os" has no attribute "sched_getaffinity"  [attr-defined]
> Found 3 errors in 2 files (checked 18 source files)```
> ```

This is likely a `mypy` version difference, try with 1.11.2

**Comment by mlopez-ntop** [comment]
> > Locally I got the following. `xdist` is from the upstream right?
> > ```
> > src\xdist\looponfail.py:169: error: Incompatible types in assignment (expression has type "ChannelFileWrite", variable has type "TextIO")  [assignment]
> > src\xdist\plugin.py:34: error: Returning Any from function declared to return "int"  [no-any-return]
> > src\xdist\plugin.py:36: error: Module "os" has no attribute "sched_getaffinity"  [attr-defined]
> > Found 3 errors in 2 files (checked 18 source files)```
> > ```
> 
> This is likely a `mypy` version difference, try with 1.11.2

A bit closer with that version
```(dev-11) C:\Users\mattlopez\Documents\pytest-xdist>python -m mypy src --ignore-missing-imports --strict
src\xdist\plugin.py:34: error: Returning Any from function declared to return "int"  [no-any-return]
src\xdist\plugin.py:36: error: Module "os" has no attribute "sched_getaffinity"  [attr-defined]
Found 2 errors in 1 file (checked 18 source files)```

**Comment by ciaconet** [comment]
> > > Locally I got the following. `xdist` is from the upstream right?
> > > ```
> > > src\xdist\looponfail.py:169: error: Incompatible types in assignment (expression has type "ChannelFileWrite", variable has type "TextIO")  [assignment]
> > > src\xdist\plugin.py:34: error: Returning Any from function declared to return "int"  [no-any-return]
> > > src\xdist\plugin.py:36: error: Module "os" has no attribute "sched_getaffinity"  [attr-defined]
> > > Found 3 errors in 2 files (checked 18 source files)```
> > > ```
> > 
> > 
> > This is likely a `mypy` version difference, try with 1.11.2
> 
> A bit closer with that version
> 
> ```
> src\xdist\plugin.py:34: error: Returning Any from function declared to return "int"  [no-any-return]
> src\xdist\plugin.py:36: error: Module "os" has no attribute "sched_getaffinity"  [attr-defined]
> Found 2 errors in 1 file (checked 18 source files)```
> ```

Those files weren't modified, I'd assume another package like `os` is at a different version, should be fine to ignore. Worst case we open another PR to our fork

**Review by mlopez-ntop — APPROVED** [review]

**Pull request merged**